### DISCUSSION TITLE
Does `Address::getLong()` provide the digital representation of an address?

### DISCUSSION THREAD

**Discussion by mandesero** [konsoletyper/teavm#979] [Q&A] [answered]
I am working with TeaVM and trying to emulate a C structure in Java using `ByteBuffer`. Here's an example of a structure in C:

```c
struct S {
    uint8_t *x;
    size_t y;
};
```

This structure would occupy 16 bytes (assuming 64-bit pointers). In Java, I attempt to create a similar structure using `ByteBuffer` and TeaVM's `Address` class:

```java
// I have a buffer
byte[] buff = new byte[1024];

// Create ByteBuffer for the structure
ByteBuffer S = ByteBuffer.allocateDirect(16);
S.order(ByteOrder.nativeOrder());

// Populate the structure
S.putLong(0, (Address.ofData(buff)).getLong()); // Address as a long
S.putLong(8, (long) 1024);                      // Buffer length
```

I am unsure whether the `getLong()` method of `Address` returns the actual digital representation of the address (pointer) in memory. Is my approach correct? If not, how can I properly achieve this?

**Comment by konsoletyper** [comment, marked as answer]
Sure, it's physical memory address, both in C and WebAssembly backends. However, note, that if you obtain `Address.ofData`, you can't rely on the fact that this is permanent address, since GC can move array into another location.

**Reply by mandesero** [reply]
> Sure, it's physical memory address, both in C and WebAssembly backends. However, note, that if you obtain `Address.ofData`, you can't rely on the fact that this is permanent address, since GC can move array into another location.

Is there a way to temporarily disable the garbage collector or somehow ensure that the memory address obtained via `Address.ofData` remains stable during operations?

**Reply by konsoletyper** [reply]
Nope, you can't disable GC. However, TeaVM GC never moves objects if their address presents directly on stack. This means that if you call native method from within Java and pass address of an array that is live-out at call site, then you can be sure that native method receives valid location. So the pattern is following: don't cache `Address`, only pass address of objects or array data to native methods called from Java and expect these methods to fill arrays.

There's also `Address.pin` method to trick liveness analysis. Note that this method does not do what usually expected from GC pinning. It just does not do anything, but instructs compiler that the local variable lives before `Address.pin`. So the pattern for getting data from native method is following:

```java
var array = new byte[desiredSize];
callWritingNativeMethod(Address.ofData(array), array.length);
// read data that were written by callWritingNativeMethod
```

Pattern for receiving data to array that is not processed by direct reference:

```java
var array = this.arrayStoredInField;
callWritingNativeMethod(Address.ofData(array), array.length);
processDataInArrayStoredAsField();
Address.pin(array);
```

Pattern for sending data to native method is following:

```java
var array = generateDataFromJava();
callReadingNativeMethod(Address.ofData(array), array.length);
```

You can also clone TeaVM sources and find all occurrences of `Address.ofData`. There are plenty methods from class library implementation that interact with POSIX and/or WASI, so you can learn used patterns and approaches.

**Reply by mandesero** [reply]
Thank you!
